File: Neural_Network_2dSLAM/ROS_data/src/dataReader.py
#! /usr/bin/env python

import rospy
from tf.transformations import euler_from_quaternion
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import Odometry
import message_filters
from time import time
import csv
import os 

dir_path = os.path.dirname(os.path.realpath(__file__))
csvFile=open('/home/heyday1006/trial/src/data_read/src/office.csv', 'w')
writer = csv.writer(csvFile, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)

def callback(laser, msg):
    # Solve all of perception here...
    quaternion=msg.pose.pose.orientation
    explicit_quat = [quaternion.x, quaternion.y, quaternion.z, quaternion.w]
    (roll,pitch,yaw) = euler_from_quaternion(explicit_quat)
    pose=(msg.pose.pose.position.x, msg.pose.pose.position.y, yaw)
    print "------------------------------------------------"
    rospy.loginfo("store info")
    print "len(range)= "+str(len(laser.ranges))
    print "[x,y,theta] = "+str(pose)
    writer.writerow([laser.ranges,pose])


def myhook():
  print "shutdown time!"

rospy.init_node("scan_values")
image_sub = message_filters.Subscriber('/robot_1/range_throttled', LaserScan)
info_sub = message_filters.Subscriber('/robot_1/odom_throttled', Odometry)

ts = message_filters.ApproximateTimeSynchronizer([image_sub, info_sub], 10,0.09)
ts.registerCallback(callback)
rospy.spin()
rospy.on_shutdown(myhook)

File: Neural_Network_2dSLAM/ROS_data/src/dataReader.cpp
#include <ros/ros.h> 
#include <sensor_msgs/LaserScan.h>
#include <nav_msgs/Odometry.h>
#include <message_filters/subscriber.h>
#include <message_filters/synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>

using namespace message_filters;

void subCallback(const sensor_msgs::LaserScan::ConstPtr& msg_1,const nav_msgs::Odometry::ConstPtr& msg_2){
    ROS_INFO_STREAM("hello");
}
int main(int argc, char** argv) {
    ros::init(argc, argv, "scan_values");
    ros::NodeHandle nh;
    message_filters::Subscriber<sensor_msgs::LaserScan> f_sub(nh, "/robot_1/base_scan", 1);
    message_filters::Subscriber<nav_msgs::Odometry> s_sub(nh, "robot_1/odom", 1);
	ROS_INFO_STREAM("hello");
    typedef sync_policies::ApproximateTime<sensor_msgs::LaserScan, nav_msgs::Odometry> MySyncPolicy;

    Synchronizer<MySyncPolicy> sync(MySyncPolicy(40), f_sub, s_sub);
    sync.registerCallback(boost::bind(&subCallback, _1, _2));
    ros::spin();
    return 0;
}
